src/ivectorbin/agglomerative-cluster.cc

```diff
@@ -0,0 +1,119 @@
+// ivectorbin/agglomerative-cluster.cc
+
+// Copyright 2016  David Snyder
+//           2017  Matthew Maciejewski
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "util/stl-utils.h"
+#include "ivector/agglomerative-bottom-up-clustering.h"
+
+int main(int argc, char *argv[]) {
+  using namespace kaldi;
+  typedef kaldi::int32 int32;
+  typedef kaldi::int64 int64;
+  try {
+    const char *usage =
+      "Cluster utterances by score, used in diarization.\n"
+      "Takes a file of scores of the form <utt1> <utt2> <score>\n"
+      "and a file spk2utt that contains the mapping from recordings\n"
+      "to the utterances within them, and outputs a list of labels\n"
+      "in the form <utt1> <label>. Clustering is done using\n"
+      "Agglomerative Hierarchical Bottom-Up Clustering with a\n"
+      "stopping-threshold score.\n"
+      "Usage: agglomerative-cluster [options] <scores-rspecifier> "
+      "<spk2utt-rspecifier> <labels-wspecifier>\n"
+      "e.g.: \n"
+      " agglomerative-cluster ark:scores ark:spk2utt \n"
+      "   ark,t:labels.txt\n";
+
+    ParseOptions po(usage);
+    std::string spk2num_rspecifier;
+    BaseFloat threshold = 0.5;
+    BaseFloat max_dist = 1.0;
+
+    po.Register("spk2num-rspecifier", &spk2num_rspecifier,
```

Comment:
Some of the places where spk is used are actually recordings. So it is better to use names like reco2utt, reco2num_spk etc.